design/api-specification-design.md

```diff
@@ -0,0 +1,83 @@
+# api-specification-design
+Personal notes about REST API design [here](https://github.com/perrymant/learning-notes/blob/master/17-rest-api-design.md)
+
+## 1. report
+The `report` request obtains a report of their account.
+
+#### Request
+| Method | URL         |
+|--------|-------------|
+| GET    | api/report  |
+
+#### Response
+- 200: a `report statement` will be sent:
+
+```json
+{
+    "data" : {
+        "report" : "---report statement-(see below)---"
+     }
+}
+```
+
+An example `report statement` for status code 200 :
```

Comment:
So let's think about the purpose of this API, and APIs in general.

Right now we're saying that we should return this nicely formatted string of transactions. This makes quite a big assumption:

> The server decides how the client should present the data.

With this approach, the client has two options:

1. Display the data exactly in the format that the API gave it. Line endings and font spacing be damned.
1. Try and parse the data out of this custom string format into more manageable data structures.

This defeats the purpose of the problem that APIs try and solve. There's no need for the API to know how the data is presented. It only needs to care about the data itself, and organise it in a logically way. Like JSON!

See if you can come up with an alternative representation of this list of transactions in JSON.

Hint: We're somewhere between 80% and 100% of the way there with the work we did around persisting the transactions to a file.